File: Python/15.3Sum.py
#找3个数，使它们的和为0，返回所有结果
class Solution(object):
    def threeSum(self, nums):
        nums.sort()         #先排序
        dct, ans = {}, []
        for i in range(0, len(nums)):
            if (i > 0 and nums[i] == nums[i-1]):
                continue     #和前面一个数一样的话，跳过
            l, r = i + 1, len(nums) - 1
            while l < r:
                sum = nums[l] + nums[r] + nums[i]
                if sum == 0:
                    ans.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]: l = l + 1  #去掉重复元素
                    while l < r and nums[r] == nums[r - 1]: r = r - 1
                    l, r = l + 1, r - 1
                elif sum < 0:
                    l = l + 1
                else:
                    r = r - 1
        return ans

s = Solution()
print(s.threeSum([-2,0,1,1,2]))

File: C++/Two Pointers/15.3Sum.cpp
/*
求数组中的三个数和为target。
*/
/*
思路：两层循环，第一层遍历值，把重复的跳过。
第二层两个指针指头尾，target为第一层的相反数。
和大于target，尾向前，小于target，头向后。
每次移动都把重复值跳过。
*/
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> result;
		int length = nums.size(), i, j, k;
		sort(nums.begin(), nums.end());
		for (k = 0; k < length - 2; ++k)
		{
			int target = -nums[k];
			i = k + 1, j = length - 1;
			while (i < j)
			{
				if (nums[i] + nums[j] == target)
					result.push_back(vector<int> {nums[k], nums[i], nums[j]});
				if (nums[i] + nums[j] < target)
				{
					++i;
					while (nums[i - 1] == nums[i]) ++i;
				}
				else
				{
					--j;
					while (j < length - 1 && nums[j + 1] == nums[j]) --j;
				}
			}
			while (nums[k] == nums[k + 1]) ++k;
		}
		return result;
    }
};
int main()
{
	Solution s;
	vector<int> nums{ 1, 0, -1, 0, -2, 2 };
	vector<vector<int>> out;
	out = s.threeSum(nums);
	for (int i = 0; i < out.size(); ++i)
	{
		for (int j = 0; j < out[i].size(); ++j)
		{
			cout << out[i][j] << "\t";
		}
		cout << endl;
	}
}

